# NLP Project 1 - Amazon Baby Products dataset

## Exploratory Data Analysis of Text data Including Visualization and Sentiment Analysis

Text data analysis is becoming easier and easier every day. Prominent programming languages like Python and R have great libraries for text data analysis. There was a time when people used to think that you need to be an expert in coding to do these types of complex tasks. But with the more developed and improved version of libraries, it is easier to perform text data analysis with just simple and beginner-level coding knowledge.
In this article, We will work on a dataset that is primarily a text dataset. The dataset contains the customer review of amazon baby products and ratings.This is a great dataset to use for machine learning as it has ratings with reviews. But this article will only focus on the exploratory data analysis

In [ ]:
.#Importing Important Libraries

In [44]:
import pandas as pd
import numpy as np
import os

In [34]:
#Code to know the current Working Directory

In [45]:
os.getcwd()

'F:\\AI & ML PRojects'

In [ ]:
#Code for Changning/Setting the new working directory

In [46]:
os.chdir('F:\AI & ML PRojects')

In [ ]:
#Code for loading the dataset

In [47]:
df= pd.read_csv("amazon_baby.csv")

In [56]:
df.head()

name  \
0                           Planetwise Flannel Wipes   
1                              Planetwise Wipe Pouch   
2                Annas Dream Full Quilt with 2 Shams   
3  Stop Pacifier Sucking without tears with Thumb...   
4  Stop Pacifier Sucking without tears with Thumb...   

                                              review  rating  
0  These flannel wipes are OK, but in my opinion ...       3  
1  it came early and was not disappointed. i love...       5  
2  Very soft and comfortable and warmer than it l...       5  
3  This is a product well worth the purchase.  I ...       5  
4  All of my kids have cried non-stop when I trie...       5

In [81]:
df.describe()

rating
count  183531.000000
mean        4.120448
std         1.285017
min         1.000000
25%         4.000000
50%         5.000000
75%         5.000000
max         5.000000

As you can see, we have the name of the products, customer reviews, and ratings in the dataset. I like to start by checking how many rows of data are there in the dataset:

In [57]:
len(df)

183531

That’s a lot of data. I decided to make the dataset smaller. Because when I tried to work with this big dataset, it takes too much time to do some visualization, calculation, and data manipulation to complete. To make it easier and less time-consuming to run the code, I decided to make the dataset smaller. One way is to simply take 5000 to 10000 or whatever number you want from the dataset.
But before doing that let’s check the value counts of the reviews for each product.

In [60]:
df['name'].value_counts()

Vulli Sophie the Giraffe Teether                                             785
Simple Wishes Hands-Free Breastpump Bra, Pink, XS-L                          562
Infant Optics DXR-5 2.4 GHz Digital Video Baby Monitor with Night Vision     561
Baby Einstein Take Along Tunes                                               547
Cloud b Twilight Constellation Night Light, Turtle                           520
                                                                            ... 
Disney Bambi Shoes for Infants -- Pink                                         1
Circle Satin Soft &amp; Cuddly Blue Snugly Baby Blanket                        1
Hoohobbers Toy Box, Primary                                                    1
Nojo Love Birds 2 Piece Canvas Wall Decor                                      1
pbnj baby SippyPal baby bottle, sippycup, and accessory holder (Blue Dot)      1
Name: name, Length: 32417, dtype: int64

You can see from the output that, there are some products that have only one review. It’s not possible to draw any conclusion from only one review. Here I am taking the products that have at least 20 reviews.

In [61]:
df1 = df[df.groupby("name")["name"].transform('size') > 20]

In [62]:
df1

name  \
153                  Fisher Price Nesting Action Vehicles   
154                  Fisher Price Nesting Action Vehicles   
155                  Fisher Price Nesting Action Vehicles   
156                  Fisher Price Nesting Action Vehicles   
157                  Fisher Price Nesting Action Vehicles   
...                                                   ...   
183511  Maxboost iPhone 5S/5 Case - Protective Snap-on...   
183512  Maxboost iPhone 5S/5 Case - Protective Snap-on...   
183513  Maxboost iPhone 5S/5 Case - Protective Snap-on...   
183514  Maxboost iPhone 5S/5 Case - Protective Snap-on...   
183515  Maxboost iPhone 5S/5 Case - Protective Snap-on...   

                                                   review  rating  
153     We bought these for our son when he turned two...       5  
154     My son loves stacking cups, so a friend recomm...       5  
155     My son Cameron just loves these great little s...       5  
156     My one year old son received these as a birthd...       5  
157     I purchased this toy for my great grandson's f...       5  
...                                                   ...     ...  
183511  I received a free sample from VIPU program for...       3  
183512  *I received a free sample from VIPU program fo...       5  
183513  I was a very surprised by the quality of this ...       5  
183514  I received a free sample from VIPU program for...       4  
183515  I love this phone case!  My  iPhone is always ...       5  

[89691 rows x 3 columns]

I check the length of the dataset again and found out that we have 89691 data now. That’s also a lot. I want only 20000 data for this demonstration. If you don’t worry about running time or if you have a higher capacity computer, please feel free to use all the data.

In [63]:
df2= df1.head(20000)

In [64]:
df2

name  \
153    Fisher Price Nesting Action Vehicles   
154    Fisher Price Nesting Action Vehicles   
155    Fisher Price Nesting Action Vehicles   
156    Fisher Price Nesting Action Vehicles   
157    Fisher Price Nesting Action Vehicles   
...                                     ...   
30710     Baby Buddy: Baby's 1st Toothbrush   
30711     Baby Buddy: Baby's 1st Toothbrush   
30712     Baby Buddy: Baby's 1st Toothbrush   
30713     Baby Buddy: Baby's 1st Toothbrush   
30714     Baby Buddy: Baby's 1st Toothbrush   

                                                  review  rating  
153    We bought these for our son when he turned two...       5  
154    My son loves stacking cups, so a friend recomm...       5  
155    My son Cameron just loves these great little s...       5  
156    My one year old son received these as a birthd...       5  
157    I purchased this toy for my great grandson's f...       5  
...                                                  ...     ...  
30710  My little one loves to brush her teeth with th...       5  
30711  I bought 2 of these-one for each of my younges...       5  
30712  I love this toothbrush. My 1 yr old started us...       5  
30713                                                NaN       3  
30714  This is an awesome infant brush and can be use...       5  

[20000 rows x 3 columns]

## Preprocessing

Some text cleaning and processing are necessary before jumping into the analysis. Preprocessing and data cleaning is a big part of data analysis. First I will make the ‘review’ column as string format. It looks like strings. But by any chance, if there are some data that are not in string format, I will simply convert the whole column as strings.

In [85]:
df['review'] = df['review'].apply(str)

In [45]:
##Reviews are the main information in this dataset. If any row is missing the review, we don’t need that row.

In [88]:
df3 = df[~df["review"].isnull()]

In [ ]:
## I still have 20000 data. That means there were no null values in the review column.

## Delete the Special Characters

In [ ]:
##Reviews may contain a lot of special characters that are not helpful for analysis. It’s good to clean them in the beginning.

In [89]:
def clean(txt):
    txt = txt.str.replace("(<br/>)", "")
    txt = txt.str.replace('(<a).*(>).*(</a>)', '')
    txt = txt.str.replace('(&amp)', '')
    txt = txt.str.replace('(&gt)', '')
    txt = txt.str.replace('(&lt)', '')
    txt = txt.str.replace('(\xa0)', ' ')  
    return txt
df['review'] = clean(df['review'])

## Converting to lower case

Converting to lower case is necessary. Otherwise, it will consider the same word with an uppercase as a different word. Like ‘me’ and ‘Me’ will be considered as different words. We don’t want that.

In [90]:
df['review1'] = df['review'].apply(lambda x: " ".join(x.lower() for x in x.split()))
df['review1'].head()

0    these flannel wipes are ok, but in my opinion ...
1    it came early and was not disappointed. i love...
2    very soft and comfortable and warmer than it l...
3    this is a product well worth the purchase. i h...
4    all of my kids have cried non-stop when i trie...
Name: review1, dtype: object

## Removing the Punctuation

This step is to remove the punctuations. Because of punctuation, a word might be treated differently than it originally is. For example ‘use’ and ‘use:’ will become different words because of punctuation.

In [92]:
df['review1'] = df['review1'].str.replace('[^\w\s]', '')
df['review1'].head()

0    these flannel wipes are ok but in my opinion n...
1    it came early and was not disappointed i love ...
2    very soft and comfortable and warmer than it l...
3    this is a product well worth the purchase i ha...
4    all of my kids have cried nonstop when i tried...
Name: review1, dtype: object